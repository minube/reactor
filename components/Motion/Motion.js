import { array, bool, node, object, oneOfType, string, number } from 'prop-types';
import React, { PureComponent } from 'react';
import { Animated, Platform, StyleSheet, View as ViewNative } from 'react-native';

const SPRING = 'spring';
const SPRING_BEZIER = 'cubic-bezier(0.175, 0.885, 0.160, 1.105)';
const TRANSFORM_PROPERTIES = ['scale', 'translateX', 'translateY'];

class Motion extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: new Animated.Value(props.useNativeDriver ? 0 : props.value),
    };
  }

  componentWillReceiveProps({
    delay = this.props.delay,
    disabled = this.props.disabled,
    duration = this.props.duration,
    type = this.props.type,
    useNativeDriver = this.props.useNativeDriver,
    value = this.props.duration || 0,
  }) {
    if (!disabled && !useNativeDriver && value !== this.props.value) {
      Animated[type](this.state.value, { toValue: value, delay, duration }).start();
    }
  }

  render() {
    const {
      props: {
        children, delay, disabled, duration, property, style, type, useNativeDriver,
      },
      state: { value },
    } = this;
    const View = !disabled && useNativeDriver ? ViewNative : Animated.View;
    const isTransform = TRANSFORM_PROPERTIES.includes(property);
    const transitionValue = useNativeDriver ? this.props.value : value;
    const transitionProperty = isTransform ? 'transform' : property;

    const styleMotion = !disabled
      ?
      StyleSheet.flatten([
        property && useNativeDriver && {
          transitionProperty,
          transitionDelay: `${delay}ms`,
          transitionDuration: `${duration}ms`,
          transitionTimingFunction: type === SPRING ? SPRING_BEZIER : undefined,
        },
        property && { [transitionProperty]: isTransform ? [{ [property]: transitionValue }] : transitionValue },
      ])
      : undefined;

    return (
      <View style={StyleSheet.flatten([style, styleMotion])}>
        { children }
      </View>
    );
  }
}

Motion.propTypes = {
  children: node,
  delay: number,
  disabled: bool,
  duration: number,
  property: string,
  style: oneOfType([array, number, object]),
  type: string,
  useNativeDriver: bool,
  value: number,
};

Motion.defaultProps = {
  children: undefined,
  delay: 0,
  disabled: false,
  duration: 500,
  property: undefined,
  style: [],
  type: SPRING,
  useNativeDriver: Platform.OS === 'web',
  value: 0,
};

export default Motion;
