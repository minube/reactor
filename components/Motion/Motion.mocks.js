import { THEME } from '../../common';

const STYLE = { backgroundColor: THEME.COLOR.PRIMARY, padding: 32 };
const STYLE_VIEW = {
  width: 128,
  height: 128,
  backgroundColor: THEME.COLOR.SECONDARY,
};

const TIMELINE = [
  { property: 'opacity', value: 0.75 },
  { property: 'scale', value: 0.5 },
  { property: 'left', value: 128 },
];

export default { STYLE, STYLE_VIEW, TIMELINE };
