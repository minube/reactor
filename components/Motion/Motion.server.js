import { createElement } from 'react';
import { View } from 'react-native';

export default ({ children, style }) => createElement(View, { style }, children);
