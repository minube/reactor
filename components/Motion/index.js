import { ENV } from '../../common';
import Default from './Motion';

const Loading = ENV.IS_SERVER ? require('./Motion.server').default : Default;

export default Loading;
