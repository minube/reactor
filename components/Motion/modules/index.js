import buildStyle from './buildStyle';

export { buildStyle }; // eslint-disable-line
