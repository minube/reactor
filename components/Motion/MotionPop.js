import { bool, node } from 'prop-types';
import { createElement, PureComponent } from 'react';

import Motion from './Motion';

class MotionPop extends PureComponent {
  static propTypes = {
    children: node.isRequired,
    visible: bool,
  };

  static defaultProps = {
    visible: false,
  };

  render() {
    const {
      props: { children, visible, ...inherit },
    } = this;

    return createElement(Motion, {
      ...inherit,
      timeline: [
        { property: 'opacity', value: visible ? 1 : 0 },
        { property: 'scale', value: visible ? 1 : 0 },
      ],
    }, children);
  }
}

export default MotionPop;
