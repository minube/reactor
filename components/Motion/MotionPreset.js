import { bool, oneOf, node } from 'prop-types';
import { createElement, PureComponent } from 'react';

import Motion from './Motion';
import PRESETS from './modules/presets';

class MotionPreset extends PureComponent {
  static propTypes = {
    children: node.isRequired,
    value: oneOf(['fade', 'pop', 'pop_fade']),
    visible: bool,
  };

  static defaultProps = {
    value: undefined,
    visible: false,
  };

  render() {
    const {
      children, value = '', visible, ...inherit
    } = this.props;

    const { IN, OUT } = PRESETS[value.toUpperCase()] || PRESETS.POP_FADE;

    return createElement(Motion, {
      ...inherit,
      pointerEvents: visible ? undefined : 'none',
      timeline: visible ? IN : OUT,
    }, children);
  }
}

export default MotionPreset;
