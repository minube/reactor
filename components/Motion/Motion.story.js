import React from 'react';
import { View } from 'react-native';
import { storiesOf } from '@storybook/react';
import { boolean, object, number, select } from '@storybook/addon-knobs/react';

import Motion from './Motion';
import MotionPreset from './MotionPreset';

const STYLE = { backgroundColor: 'rgba(0,255,0,0.25)', padding: 32 };
const STYLE_VIEW = {
  width: 128,
  height: 128,
  backgroundColor: 'pink',
};

const PROPERTIES = ['top', 'left', 'right', 'bottom', 'opacity', 'scale', 'translateX', 'translateY'];

const timeline = [
  { property: 'opacity', value: 0.75 },
  { property: 'scale', value: 0.5 },
  { property: 'left', value: 128 },
];

storiesOf('✅ Motion', module)
  .addWithJSX('default', () => (
    <Motion timeline={timeline}>
      <View style={STYLE_VIEW} />
    </Motion>
  ))
  .addWithJSX('preset', () => (
    <MotionPreset visible={boolean('visible', false)}>
      <View style={STYLE_VIEW} />
    </MotionPreset>
  ))
  .addWithJSX('style', () => (
    <Motion timeline={timeline} style={STYLE}>
      <View style={STYLE_VIEW} />
    </Motion>
  ))
  .addWithJSX('🏀 Playground', () => (
    <Motion
      delay={number('delay', 0)}
      disabled={boolean('disabled', false)}
      duration={number('duration', 500)}
      style={object('style', STYLE)}
      timeline={object('timeline', timeline)}
      type={select('type', ['spring', 'timing'], 'spring')}
      useNativeDriver={boolean('useNativeDriver', true)}
    >
      <View style={STYLE_VIEW} />
    </Motion>
  ));
