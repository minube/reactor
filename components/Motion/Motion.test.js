import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { View } from 'react-native';

import Motion from './Motion';

describe('<Motion>', () => {
  it('renders', () => {
    const tree = renderer.create(<Motion />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {property}', () => {
    const tree = renderer.create(<Motion property="left" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {type}', () => {
    const tree = renderer.create(<Motion type="timming" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {useNativeDriver}', () => {
    const tree = renderer.create(<Motion useNativeDriver />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {delay}', () => {
    const tree = renderer.create(<Motion delay={1000} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {value}', () => {
    const tree = renderer.create(<Motion property="left" value={128} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {children}', () => {
    const tree = renderer.create(<Motion><View /></Motion>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders in DOM environment', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Motion />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
});

