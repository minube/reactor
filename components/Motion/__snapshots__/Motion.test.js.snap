// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Motion> renders 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
/>
`;

exports[`<Motion> when {children} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "left": 0,
    }
  }
>
  <View />
</View>
`;

exports[`<Motion> when {delay} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "left": 0,
    }
  }
/>
`;

exports[`<Motion> when {disabled} 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
/>
`;

exports[`<Motion> when {property} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "left": 0,
    }
  }
/>
`;

exports[`<Motion> when {type} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "left": 0,
    }
  }
/>
`;

exports[`<Motion> when {useNativeDriver} 1`] = `
<View
  style={
    Object {
      "left": 0,
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "left",
      "transitionTimingFunction": "cubic-bezier(0.175, 0.885, 0.160, 1.105)",
    }
  }
/>
`;

exports[`<Motion> when {value} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "left": 128,
    }
  }
/>
`;
